File: LeetCode/LeetCode/Problems/6/Problem63.cs
namespace LeetCode.Problems
{
    public class Problem63
    {
        public int UniquePathsWithObstacles(int[][] obstacleGrid)
        {
            var rows = obstacleGrid.Length;
            var cols = obstacleGrid[0].Length;
            var dp = new int[cols];
            dp[0] = 1;

            for (var i = 0; i < rows; i++)
            {
                for (var j = 0; j < cols; j++)
                {
                    if (obstacleGrid[i][j] == 1)
                    {
                        dp[j] = 0;
                    }
                    else if (j > 0)
                    {
                        dp[j] += dp[j - 1];
                    }
                }
            }

            return dp[cols - 1];
        }
    }
}

File: leetcode_rust/src/problems/problem_63.rs
pub fn unique_paths_with_obstacles(mut obstacle_grid: Vec<Vec<i32>>) -> i32 {
    let m = obstacle_grid.len();
    let n = obstacle_grid[0].len();

    if obstacle_grid[0][0] == 1 {
        return 0;
    }

    obstacle_grid[0][0] = 1;
    for i in 0..m {
        for j in 0..n {
            if obstacle_grid[i][j] == 1 && (i != 0 || j != 0) {
                obstacle_grid[i][j] = 0;
                continue;
            }

            if i > 0 && obstacle_grid[i - 1][j] > 0 {
                obstacle_grid[i][j] += obstacle_grid[i - 1][j];
            }
            if j > 0 && obstacle_grid[i][j - 1] > 0 {
                obstacle_grid[i][j] += obstacle_grid[i][j - 1];
            }
        }
    }

    obstacle_grid[m][n]
}
